### PR TITLE
OpenTelemetryTracingProvider: Allow to configure resource detectors

### LINKED ISSUE #48885: Support OpenTelemetry resource attributes in traces
(This is used to request new product features, please visit <https://discuss.istio.io> for questions on using Istio)

**Describe the feature request**

Envoy supports resource attributes through the standard `OTEL_RESOURCE_ATTRIBUTES` environment variable, however it has to be enabled via: https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/tracers/opentelemetry/resource_detectors/v3/environment_resource_detector.proto#extension-envoy-tracers-opentelemetry-resource-detectors-environment

Ideally istio would support this in addition to tags. Resource attributes are critical to be able to set the `environment` that a service lives in.

**Describe alternatives you've considered**
There are no alternatives, you cannot change this property on istio-proxies. Perhaps it would be possible with an EnvoyFilter?

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[ ] Dual Stack
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[x] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[x] Multi Cluster
[x] Virtual Machine
[x] Multi Control Plane

### PR DESCRIPTION
**Pull request by joaopgrassi**
**Please provide a description of this PR:**
This PR uses the new resource detectors API config (https://github.com/istio/api/pull/3002) and allows configuring OpenTelemetry [resource detectors](https://opentelemetry.io/docs/languages/js/resources/#process--environment-resource-detection). 

Resource detectors are an important part of observability, allowing to correlate telemetry with things such as hosts, processes, cloud resources and much more. This PR adds the possibility to configure two resource detectors: 

- [Environment](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/tracers/opentelemetry/resource_detectors/v3/environment_resource_detector.proto#extension-envoy-tracers-opentelemetry-resource-detectors-environment)
- [Dynatrace](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/tracers/opentelemetry/resource_detectors/v3/dynatrace_resource_detector.proto#extension-envoy-tracers-opentelemetry-resource-detectors-dynatrace)

Closes #48885

I tested this locally with both resource detectors enabled and I confirmed the attributes are all added to the exported telemetry.

### CODE REVIEW

**Comment by joaopgrassi** [comment]
The tests are failing because when configuring the `environment` resource detector, Envoy fails to start when the env variable `OTEL_RESOURCE_ATTRIBUTES` is not defined. I'm looking if I can find a way to add some but would appreciate some help.

**Review by zirain — APPROVED** [review]

File: tests/integration/telemetry/tracing/otelcollector/tracing_test.go [hunk 1]
```diff
@@ -187,6 +184,13 @@ meshConfig:
         headers:
           - name: "some-header"
             value: "some-value"
+  - name: test-otel-res-detectors
+    opentelemetry:
+      service: opentelemetry-collector.istio-system.svc.cluster.local
+      port: 4317
+      resource_detectors:
+        environment: {}
+        dynatrace: {}
```
Comment (zirain) [inline tests/integration/telemetry/tracing/otelcollector/tracing_test.go:193]: is this line meaningful?

File: pilot/pkg/networking/core/v1alpha3/tracing.go [hunk 2]
```diff
@@ -314,7 +312,26 @@ func otelConfig(serviceName, hostname, cluster string, otelProvider *meshconfig.
...
+		res := []*core.TypedExtensionConfig{}
+		rd := otelProvider.ResourceDetectors
+
+		if rd.Environment != nil {
+			res = append(res, &core.TypedExtensionConfig{
+				Name:        "envoy.tracers.opentelemetry.resource_detectors.environment",
+				TypedConfig: protoconv.MessageToAny(&resourcedetectors.EnvironmentResourceDetectorConfig{}),
+			})
+		}
+		if rd.Dynatrace != nil {
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/tracing.go:326]: are multiple supported?

File: tests/integration/telemetry/tracing/otelcollector/testdata/otel-tracing-res-detectors.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+apiVersion: telemetry.istio.io/v1alpha1
+kind: Telemetry
+metadata:
+  name: otel-tracing
+spec:
+  tracing:
+  - providers:
+    - name: test-otel-res-detectors
+    randomSamplingPercentage: 100.0
```
Comment (howardjohn) [inline tests/integration/telemetry/tracing/otelcollector/testdata/otel-tracing-res-detectors.yaml:9]: In the PR we discussed the case of a user setting sampling to not 100% we would override it. I don't see that implemented, though

Comment (joaopgrassi) [reply tests/integration/telemetry/tracing/otelcollector/testdata/otel-tracing-res-detectors.yaml:9]: This PR is just about resource detectors. Sampling will come in a separated PR.

File: pilot/pkg/networking/core/v1alpha3/tracing.go (hunk 2 above)
Comment (joaopgrassi) [reply pilot/pkg/networking/core/v1alpha3/tracing.go:326]: No, it doesn't make sense to support multiple of the same as they would simply override each other (the data it produces)

File: tests/integration/telemetry/tracing/otelcollector/testdata/otel-tracing-res-detectors.yaml (hunk 3 above)
Comment (howardjohn) [reply tests/integration/telemetry/tracing/otelcollector/testdata/otel-tracing-res-detectors.yaml:9]: oops, thanks! got them mixed up

File: pilot/pkg/networking/core/v1alpha3/tracing.go (hunk 2 above)
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/tracing.go:326]: but we are appending here, so someone could set dynatrace and environment. Your comment above sounds like that is not expected?

Comment (zirain) [reply pilot/pkg/networking/core/v1alpha3/tracing.go:326]: it's expected, but should avoid multiple detectors in same type(either dynatrace or environment).

**Comment by zirain** [comment]
> The tests are failing because when configuring the `environment` resource detector, Envoy fails to start when the env variable `OTEL_RESOURCE_ATTRIBUTES` is not defined. I'm looking if I can find a way to add some but would appreciate some help.
> 
> ```
> ner rejected: Error adding/updating listener(s) 0.0.0.0_9080: Environment variable doesn't exist: OTEL_RESOURCE_ATTRIBUTES
> 0.0.0.0_4317: Environment variable doesn't exist: OTEL_RESOURCE_ATTRIBUTES
> 0.0.0.0_15010: Environment variable doesn't exist: OTEL_RESOURCE_ATTRIBUTES
> 10.98.50.165_15021: Environment variable doesn't exist: OTEL_RESOURCE_ATTRIBUTES
> 0.0.0.0_15014: Environment variable doesn't exist: OTEL_RESOURCE_ATTRIBUTES
> 0.0.0.0_80: Environment variable doesn't exist: OTEL_RESOURCE_ATTRIBUTES
> 10.96.0.10_9153: Environment variable doesn't exist: OTEL_RESOURCE_ATTRIBUTES
> 0.0.0.0_55678: Environment variable doesn't exist: OTEL_RESOURCE_ATTRIBUTES
> 0.0.0.0_4318: Environment variable doesn't exist: OTEL_RESOURCE_ATTRIBUTES
> virtualInbound: Environment variable doesn't exist: OTEL_RESOURCE_ATTRIBUTES
>         thread=14
> 2024-02-05T12:46:00.634678Z     warn    Envoy proxy is NOT ready: config received from XDS server, but was rejected: cds updates: 1 successful, 0 rejected; lds updates: 0 successful, 1 rejected
> ```

maybe envoy should just warning instead of rejecting the config.

File: tests/integration/telemetry/tracing/otelcollector/testdata/otel-tracing-res-detectors.yaml (hunk 3 above)
Comment (joaopgrassi) [reply tests/integration/telemetry/tracing/otelcollector/testdata/otel-tracing-res-detectors.yaml:9]: No prob!

**Comment by joaopgrassi** [comment]
> maybe envoy should just warning instead of rejecting the config.

Yes, that was my proposal as well but the Envoy folks preferred that if there are  problems during "configuration phase", Envoy shouldn't not start up. The argument was that if that failed and Envoy would just continue silently, it would be hard for users to know that the resource detector is not working properly. 

The Dynatrace detector is different (if it fails to detect, Envoy is not blocked), because if it fails for some reason, Dynatrace customers will know inside the product, on their deployment status page.

File: pilot/pkg/networking/core/v1alpha3/tracing.go (hunk 2 above)
Comment (joaopgrassi) [reply pilot/pkg/networking/core/v1alpha3/tracing.go:326]: > but we are appending here, so someone could set dynatrace and environment. Your comment above sounds like that is not expected?

Ah right, sorry. Yes it is totally possible to have multiple resource detectors configured - AWS, Environment, Dynatrace. All of them detect different attributes and merge all into a final Resource. 

What I meant is that it doesn't make sense to configure the same detector twice, as they would just override each other.

File: tests/integration/telemetry/tracing/otelcollector/tracing_test.go (hunk 1 above)
Comment (joaopgrassi) [reply tests/integration/telemetry/tracing/otelcollector/tracing_test.go:193]: I wanted to have a test that has all detectors, to make sure things work still. But if you prefer I can remove the Dynatrace one and leave only the environment.

File: pilot/pkg/networking/core/v1alpha3/tracing.go [hunk 4]
```diff
@@ -314,7 +312,26 @@ func otelConfig(serviceName, hostname, cluster string, otelProvider *meshconfig.
...
 	}
 
-	oc.ServiceName = serviceName
+	// Add configured resource detectors
+	if otelProvider.ResourceDetectors != nil {
+		res := []*core.TypedExtensionConfig{}
+		rd := otelProvider.ResourceDetectors
+
+		if rd.Environment != nil {
+			res = append(res, &core.TypedExtensionConfig{
```
[the remark this answers is not in the document]
Comment (joaopgrassi) [reply pilot/pkg/networking/core/v1alpha3/tracing.go:321]: Sorry, I'm not really familiar with the code base, what exactly should I add there? Do you have maybe an example?

**Comment by zirain** [comment]
> > maybe envoy should just warning instead of rejecting the config.
> 
> Yes, that was my proposal as well but the Envoy folks preferred that if there are problems during "configuration phase", Envoy [shouldn't not start up](https://github.com/envoyproxy/envoy/pull/29547#discussion_r1324548685). The argument was that if that failed and Envoy would just continue silently, it would be hard for users to know that the resource detector is not working properly.
> 
> The Dynatrace detector is different (if it fails to detect, Envoy is not blocked), because if it fails for some reason, Dynatrace customers will know inside the product, on their deployment status page.

make sense

**Comment by howardjohn** [comment]
> > > maybe envoy should just warning instead of rejecting the config.
> > 
> > 
> > Yes, that was my proposal as well but the Envoy folks preferred that if there are problems during "configuration phase", Envoy [shouldn't not start up](https://github.com/envoyproxy/envoy/pull/29547#discussion_r1324548685). The argument was that if that failed and Envoy would just continue silently, it would be hard for users to know that the resource detector is not working properly.
> > The Dynatrace detector is different (if it fails to detect, Envoy is not blocked), because if it fails for some reason, Dynatrace customers will know inside the product, on their deployment status page.
> 
> make sense

I do not agree it makes sense :slightly_smiling_face: . This turns this feature into a cluster-wide outage generator. I really think this needs more consideration

**Review by howardjohn — CHANGES_REQUESTED** [review]
We need to make sure environment detector cannot produce outages. Maybe changing Envoy, or we could add some detection in Istio (strongly prefer not to do this one)

**Comment by howardjohn** [comment]
Otel proper SDKs do not fail if OTEL_RESOURCE_ATTRIBUTES is unset, I think this basically violates OTEL spec.

**Comment by joaopgrassi** [comment]
> Otel proper SDKs do not fail if OTEL_RESOURCE_ATTRIBUTES is unset, I think this basically violates OTEL spec.

@howardjohn  In general I agree with you and that is what I tried to push and gave examples (OTel collector also does not fail), but unfortunately my intention didn't make it through. 

About violating the OTel spec, unfortunately it is not that simple:

The OTel [spec for resource detectors](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.29.0/specification/resource/sdk.md#detecting-resource-information-from-the-environment) says:

> Resource detection logic is expected to complete quickly since this code will be run during application initialization. Errors should be handled as specified in the [Error Handling principles](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.29.0/specification/error-handling.md#basic-error-handling-principles). Note the **failure to detect any resource information MUST NOT be considered an error**, whereas an error that occurs during an attempt to detect resource information SHOULD be considered an error.

The emphasis is mine, of what is pertinent to the discussion here. If the user configures Istio to use the env resource detector, but doesn't have the variable set, it shouldn't throw errors - as that is error during detecting resource information. 

BUT: the [Error Handling principles](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.29.0/specification/error-handling.md#basic-error-handling-principles) it links says:

> The API or SDK MAY fail fast and cause the application to fail on initialization, e.g. **because of a bad user config** or environment, but MUST NOT cause the application to fail later at run time, e.g. due to dynamic config settings received from the Collector.

Which then favors what Envoy is doing - a user configuring Envoy to have the environment resource detector but NOT setting the environment variable can be considered as "bad user config" so it's not actually violating the spec. 

Not sure how I can proceed here. I can try to open a new PR in Envoy, trying to change the behavior of the detector and I would much appreciate help from you folks to maybe have more weight on the decision.

**Comment by howardjohn** [comment]
I opened https://github.com/envoyproxy/envoy/issues/32225

**Comment by joaopgrassi** [comment]
> I opened [envoyproxy/envoy#32225](https://github.com/envoyproxy/envoy/issues/32225)

Thanks! I opened the PR, let's see if we can get it patched somehow.

File: pilot/pkg/networking/core/v1alpha3/tracing.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (joaopgrassi) [reply pilot/pkg/networking/core/v1alpha3/tracing.go]: Did in [1cda858](https://github.com/istio/istio/pull/49204/commits/1cda858d0633f8464570f4f685d4f75e1620126e). Please let me know if that's what you had in mind.

**Comment by joaopgrassi** [comment]
> I opened [envoyproxy/envoy#32225](https://github.com/envoyproxy/envoy/issues/32225)

The PR that fixes it is merged in Envoy. 

I reverted the change that added the env variable to the test. So it should fail now until Istio has the new Envoy version with the fix.

File: pilot/pkg/networking/core/v1alpha3/tracing.go (hunk 2 above)
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/tracing.go:323]: ACK, thanks!

**Review by howardjohn — APPROVED** [review]
Thanks!!

**Comment by joaopgrassi** [comment]
Istio Proxy is still not up-to date with my commit in Envoy, but soon 🔜 . https://github.com/istio/proxy/commit/4d0b802082dd36cd1a5597ca24ed1fefbcb55f33

**Pull request merged**